crates/pink/src/local_cache.rs

```diff
@@ -0,0 +1,220 @@
+//! The LocalCache provides a local KV cache for contracts to do some offchain computation.
+//! When we say local, it means that the data stored in the cache is different in different
+//! machines of the same contract. And the data might loss when the pruntime restart or caused
+//! by some kind of cache expiring machanism.
+
+use alloc::borrow::Cow;
+use once_cell::sync::Lazy;
+use std::{collections::HashMap, sync::RwLock, time::Instant};
+
+pub use pink_extension::chain_extension::StorageQuotaExceeded;
+
+pub static GLOBAL_CACHE: Lazy<RwLock<LocalCache>> = Lazy::new(Default::default);
+
+#[derive(Default, Debug)]
+struct Storage {
+    // Sum of the size of all the keys and values.
+    size: usize,
+    kvs: HashMap<Vec<u8>, StorageValue>,
+}
+
+#[derive(Debug)]
+struct StorageValue {
+    // Expiration time in seconds since the first call to `now`.
+    expire_at: u64,
+    value: Vec<u8>,
+}
+
+#[derive(Debug)]
+pub struct LocalCache {
+    // Number of set ops between two GC ops.
+    gc_interval: u64,
+    // Accumulated number of set ops since last GC.
+    sets_since_last_gc: u64,
+    // Default expiration time in seconds.
+    default_value_lifetime: u64,
+    max_cache_size_per_contract: usize,
+    storages: HashMap<Vec<u8>, Storage>,
+}
+
+impl Default for LocalCache {
+    fn default() -> Self {
+        Self {
+            gc_interval: 1000,
+            sets_since_last_gc: 0,
+            default_value_lifetime: 3600 * 24 * 7, // 1 week
+            max_cache_size_per_contract: 10 * 1024 * 1024, // 10MB
+            storages: Default::default(),
+        }
+    }
+}
+
+impl LocalCache {
+    fn maybe_clear_expired(&mut self) {
+        self.sets_since_last_gc += 1;
+        if self.sets_since_last_gc == self.gc_interval {
+            self.sets_since_last_gc = 0;
+            let now = now();
+            self.storages.values_mut().for_each(|storage| {
+                let storage_size = &mut storage.size;
+                storage.kvs.retain(|k, v| {
+                    if v.expire_at > now {
+                        true
+                    } else {
+                        *storage_size -= v.value.len() + k.len();
+                        false
+                    }
+                });
+            });
+        }
+    }
+
+    pub fn get(&self, id: &[u8], key: &[u8]) -> Option<Vec<u8>> {
+        Some(self.storages.get(id)?.kvs.get(key)?.value.to_owned())
+    }
+
+    pub fn set(
+        &mut self,
+        id: Cow<[u8]>,
+        key: Cow<[u8]>,
+        value: Cow<[u8]>,
+    ) -> Result<(), StorageQuotaExceeded> {
+        self.maybe_clear_expired();
```

Comment:
Not familiar with the common practice, but should we trigger the expiration check when reading / removing / setting expiration of the kv entry?

Reply:
There might be not much common practice. Common memory GCs trigger it in allocation api. But that's their technical limitation because they has no ablity to trigger GC at read time. But we have. So more opportunities to trigger the check might make the memory usage more efficient.

Reply:
Added checks except in the get method because of the immutable `&self`.